make javascript fails with SpiderMonkey 60 on Debian Buster

@jiangphcn please see:

https://ci-couchdb.apache.org/blue/organizations/jenkins/jenkins-cm1%2FPullRequests/detail/PR-2367/44/pipeline/62

This is the new CI setup, which is proposed to use the precompiled Debian Buster libmozjs60 that the OS provides (package libmozjs60-dev).

The error we're getting is:

```
[2020-01-04T22:31:34.226Z] test/javascript/tests/view_include_docs.js                     
[2020-01-04T22:31:34.226Z]     Error: Compilation of the map function in the 'with_id' view failed: (new String("out of memory")) (function(doc) {if(doc.link_id) { var value = {'_id':doc.link_id}; if (doc.link_rev) {value._rev = doc.link_rev}; emit(doc._id, value);}};)
[2020-01-04T22:31:34.226Z] Trace back (most recent call first):
[2020-01-04T22:31:34.226Z]     
[2020-01-04T22:31:34.226Z]  551:14: test/javascript/couch.js
[2020-01-04T22:31:34.226Z]       CouchError
[2020-01-04T22:31:34.226Z]  514:12: test/javascript/couch.js
[2020-01-04T22:31:34.226Z]       CouchDB.maybeThrowError
[2020-01-04T22:31:34.226Z]  58:5: test/javascript/couch.js
[2020-01-04T22:31:34.226Z]       CouchDB/this.save
[2020-01-04T22:31:34.226Z]  41:5: test/javascript/tests/view_include_docs.js
[2020-01-04T22:31:34.226Z]       couchTests.view_include_docs
[2020-01-04T22:31:34.226Z]  45:5: test/javascript/cli_runner.js
[2020-01-04T22:31:34.226Z]       runTest
[2020-01-04T22:31:34.226Z]  56:1: test/javascript/cli_runner.js
[2020-01-04T22:31:34.226Z]       
[2020-01-04T22:31:34.226Z] fail
[2020-01-04T22:31:34.226Z] =======================================================
[2020-01-04T22:31:34.226Z] JavaScript tests complete.
[2020-01-04T22:31:34.226Z]   Failed: 1.  Skipped or passed: 76.
```

Can you have a look?  /cc @janl

hey @davisp , @wohali  , @janl , after some investigation, I highly suspect that this `"out of memory"` is related to `infinite_loop` test in https://github.com/apache/couchdb/blob/master/test/javascript/tests/view_errors.js#L154
```
      var designDoc3 = {
        _id:"_design/infinite",
        language: "javascript",
        views: {
          "infinite_loop" :{map:"function(doc) {while(true){emit(doc,doc);}};"}
        }
      };
      T(db.save(designDoc3).ok);

      try {
          db.view("infinite/infinite_loop");
          T(0 == 1);
      } catch(e) {
          T(e.error == "os_process_error" || e.error == "unnamed_error");
      }
```

The sequence of test suites follows `view_errors.js` and then `view_include_docs.js`. In some case, `couchjs` is shared among different test suites. After `view_errors` bumped memory, there is no room for `view_include_docs.js` to allocate when the shared `couchjs` is used.

Today, I ever made several test in docker image
1. run `make javascript suites="view_include_docs"` more than 200 times in the loop, and didn't hit out-of-memory issue
2. run `make javascript suites="view_errors  view_include_docs"`, and i can get out-of-memory sometime, like

```
root@0010847bf784:/couchdb# make javascript suites="view_errors  view_include_docs"
test/javascript/tests/view_errors.js         pass
test/javascript/tests/view_include_docs.js
    Error: (new String("out of memory")) (function (doc) { if (typeof doc.integer === 'number') {emit(doc.integer, doc.integer)};})
Trace back (most recent call first):

 551:14: test/javascript/couch.js
      CouchError
 514:12: test/javascript/couch.js
      CouchDB.maybeThrowError
 202:5: test/javascript/couch.js
      CouchDB/this.view
 43:14: test/javascript/tests/view_include_docs.js
      couchTests.view_include_docs
 45:5: test/javascript/cli_runner.js
      runTest
 56:1: test/javascript/cli_runner.js

fail
=======================================================
JavaScript tests complete.
  Failed: 1.  Skipped or passed: 1.
make: *** [Makefile:279: javascript] Error 1
```

3. run `make javascript suites="view_errors  view_include_docs"` after commenting out below in `view_errors.js`. I didn't see out-of-memory when running more than 100 times 
```
      try {
          db.view("infinite/infinite_loop");
          T(0 == 1);
      } catch(e) {
          //T(e.error == "os_process_error" || e.error == "unnamed_error");
      }
```

Could we give me a try in Jenkins server to disable infinite_loop temporarily?

--- a/src/couch/priv/couch_js/60/main.cpp
+++ b/src/couch/priv/couch_js/60/main.cpp
@@ -420,6 +420,7 @@ main(int argc, const char* argv[])
         return 1;
 
     JS::SetWarningReporter(cx, couch_error);
+    JS::SetOutOfMemoryCallback(cx, couch_oom, NULL);
     JS_SetContextPrivate(cx, args);
     JS_SetSecurityCallbacks(cx, &security_callbacks);
 

--- a/src/couch/priv/couch_js/60/util.cpp
+++ b/src/couch/priv/couch_js/60/util.cpp
@@ -309,6 +309,14 @@ couch_error(JSContext* cx, JSErrorReport* report)
 }
 
 
+void
+couch_oom(JSContext* cx, void* data)
+{
+    fprintf(stderr, "out of memory\n");
+    exit(1);
+}
+
+
 bool
 couch_load_funcs(JSContext* cx, JS::HandleObject obj, JSFunctionSpec* funcs)
 {

--- a/src/couch/priv/couch_js/60/util.h
+++ b/src/couch/priv/couch_js/60/util.h
@@ -35,6 +35,7 @@ JSString* couch_readline(JSContext* cx, FILE* fp);
 size_t couch_readfile(const char* file, char** outbuf_p);
 void couch_print(JSContext* cx, unsigned int argc, JS::CallArgs argv);
 void couch_error(JSContext* cx, JSErrorReport* report);
+void couch_oom(JSContext* cx, void* data);
 bool couch_load_funcs(JSContext* cx, JS::HandleObject obj, JSFunctionSpec* funcs);
 
 

--- a/src/couch/test/eunit/couch_js_tests.erl
+++ b/src/couch/test/eunit/couch_js_tests.erl
@@ -1,0 +1,45 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(couch_js_tests).
+-include_lib("eunit/include/eunit.hrl").
+
+
+-define(FUNC, <<
+  "function(doc) {\n"
+  "  var val = \"0123456789ABCDEF\";\n"
+  "  while(true) {emit(val, val);}\n"
+  "}\n"
+>>).
+
+
+couch_js_test_() ->
+    {
+        "Test couchjs",
+        {
+            setup,
+            fun test_util:start_couch/0,
+            fun test_util:stop_couch/1,
+            [
+                fun should_exit_on_oom/0
+            ]
+        }
+    }.
+
+
+should_exit_on_oom() ->
+    Proc = couch_query_servers:get_os_process(<<"javascript">>),
+    true = couch_query_servers:proc_prompt(Proc, [<<"add_fun">>, ?FUNC]),
+    ?assertThrow(
+            {os_process_error, {exit_status, 1}},
+            couch_query_servers:proc_prompt(Proc, [<<"map_doc">>, <<"{}">>])
+        ).

--- a/test/javascript/tests/auth_cache.js
+++ b/test/javascript/tests/auth_cache.js
@@ -10,8 +10,8 @@
 // License for the specific language governing permissions and limitations under
 // the License.
 
+couchTests.elixir = true;
 couchTests.auth_cache = function(debug) {
-  return console.log('done in test/elixir/test/auth_cache_test.exs');
   if (debug) debugger;
 
   // Simple secret key generator

--- a/test/javascript/tests/cookie_auth.js
+++ b/test/javascript/tests/cookie_auth.js
@@ -10,9 +10,9 @@
 // License for the specific language governing permissions and limitations under
 // the License.
 
+couchTests.elixir = true;
 couchTests.cookie_auth = function(debug) {
   // This tests cookie-based authentication.
-  return console.log('done in test/elixir/test/cookie_auth_test.exs');
 
   var db_name = get_random_db_name();
   var db = new CouchDB(db_name, {"X-Couch-Full-Commit":"false"});

--- a/test/javascript/tests/users_db.js
+++ b/test/javascript/tests/users_db.js
@@ -10,8 +10,8 @@
 // License for the specific language governing permissions and limitations under
 // the License.
 
+couchTests.elixir = true;
 couchTests.users_db = function(debug) {
-  return console.log('done in test/elixir/test/users_db_test.exs');
 
   // This tests the users db, especially validations
   // this should also test that you can log into the couch

--- a/test/javascript/tests/utf8.js
+++ b/test/javascript/tests/utf8.js
@@ -10,8 +10,8 @@
 // License for the specific language governing permissions and limitations under
 // the License.
 
+couchTests.elixir = true;
 couchTests.utf8 = function(debug) {
-  return console.log('done in test/elixir/test/utf8_test.exs');
   var db_name = get_random_db_name();
   var db = new CouchDB(db_name, {"X-Couch-Full-Commit":"false"});
   db.createDb();

--- a/test/javascript/tests/view_errors.js
+++ b/test/javascript/tests/view_errors.js
@@ -154,7 +154,7 @@ couchTests.view_errors = function(debug) {
           db.view("infinite/infinite_loop");
           T(0 == 1);
       } catch(e) {
-          T(e.error == "os_process_error" || e.error == "unnamed_error");
+          T(e.error == "os_process_error");
       }
 
       // Check error responses for invalid multi-get bodies.

